internal/webhook/management_webhook.go

```diff
@@ -55,21 +58,109 @@ func (*ManagementValidator) ValidateCreate(_ context.Context, _ runtime.Object)
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
-func (v *ManagementValidator) ValidateUpdate(ctx context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+func (v *ManagementValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
 	const invalidMgmtMsg = "the Management is invalid"
 
-	mgmt, ok := newObj.(*hmcv1alpha1.Management)
+	newMgmt, ok := newObj.(*hmcv1alpha1.Management)
 	if !ok {
 		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected Management but got a %T", newObj))
 	}
 
+	oldMgmt, ok := oldObj.(*hmcv1alpha1.Management)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected Management but got a %T", oldObj))
+	}
+
+	if err := checkComponentsRemoval(ctx, v.Client, oldMgmt, newMgmt); err != nil {
+		return admission.Warnings{"Some of the providers cannot be removed"},
+			apierrors.NewInvalid(newMgmt.GroupVersionKind().GroupKind(), newMgmt.Name, field.ErrorList{
+				field.Forbidden(field.NewPath("spec", "providers"), err.Error()),
+			})
+	}
+
+	incompatibleContracts, err := getIncompatibleContracts(ctx, v, newMgmt)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %v", invalidMgmtMsg, err)
+	}
+
+	if incompatibleContracts != "" {
+		return admission.Warnings{"The Management object has incompatible CAPI contract versions in ProviderTemplates"}, fmt.Errorf("%s: %s", invalidMgmtMsg, incompatibleContracts)
+	}
+
+	return nil, nil
+}
+
+func checkComponentsRemoval(ctx context.Context, cl client.Client, oldMgmt, newMgmt *hmcv1alpha1.Management) error {
+	removedComponentNames := []string{}
```

Comment:
nit: change to a map to not iterate over a slice in the loop

Reply:
arrays are more performant on such small datasets even when only misses are being benchmarked 

```
goos: darwin
goarch: arm64
pkg: github.com/zerospiel/playground
cpu: Apple M1 Pro
BenchmarkStrs/arrays-10         	44406596	        28.77 ns/op	       0 B/op	       0 allocs/op
BenchmarkStrs/maps-10           	12531898	        97.03 ns/op	       0 B/op	       0 allocs/op
```